### PR TITLE
feat: Add Synapse API description project

### PR DESCRIPTION
**Pull request by tschaffter**
Contributes to #1670

## Changelog

- Create the project `synapse-api-description`
- Define the path `/challenge` (adapted from [synapse-rest-openapi](https://github.com/Sage-Bionetworks/synapse-rest-openapi))

## Notes

- Setting up this project from the template provided by `schematic-api-description` took me less than 30 min.
- This PR is very minimal and just aims to get started with the generation of an API client.

### Build the API description

```console
vscode@ab1a700c294d:/workspaces/sage-monorepo$ nx build synapse-api-description

> nx run synapse-api-description:build

bundling src/openapi.yaml...
📦 Created a bundle for src/openapi.yaml at build/openapi.yaml 126ms.
```

### Format the files with Prettier

The files are also formatted when saving them in VS Code.

```console
vscode@ab1a700c294d:/workspaces/sage-monorepo$ nx run synapse-api-description:format

> nx run synapse-api-description:format

libs/synapse/api-description/docs/index.html 114ms
libs/synapse/api-description/openapi-lint-config.yaml 13ms
libs/synapse/api-description/project.json 68ms
libs/synapse/api-description/README.md 40ms
libs/synapse/api-description/spectral.yaml 23ms
libs/synapse/api-description/src/components/headers/ExpiresAfter.yaml 4ms
libs/synapse/api-description/src/components/README.md 16ms
libs/synapse/api-description/src/components/responses/BadRequest.yaml 3ms
libs/synapse/api-description/src/components/responses/Conflict.yaml 4ms
libs/synapse/api-description/src/components/responses/InternalServerError.yaml 4ms
libs/synapse/api-description/src/components/responses/NotFound.yaml 6ms
libs/synapse/api-description/src/components/responses/Unauthorized.yaml 5ms
libs/synapse/api-description/src/components/schemas/BasicError.yaml 9ms
libs/synapse/api-description/src/components/schemas/ChallengePagedResults.yaml 6ms
libs/synapse/api-description/src/components/schemas/PageMetadata.yaml 14ms
libs/synapse/api-description/src/openapi.yaml 11ms
libs/synapse/api-description/src/paths/challenge.yaml 20ms
libs/synapse/api-description/src/paths/README.md 42ms
```

### Validate the API description with IBM OpenAPI Validator

```console
vscode@ab1a700c294d:/workspaces/sage-monorepo$ nx run synapse-api-description:lint

> nx run synapse-api-description:lint  [existing outputs match the cache, left as is]

IBM OpenAPI Validator (validator: 1.0.7; ruleset: 1.0.4), @Copyright IBM Corporation 2017, 2023.

Validation Results for build/openapi.yaml:

'lodash/isPlainObject' is imported by lodash/isPlainObject?commonjs-external, but could not be resolved – treating it as an external dependency
'lodash/isEqual' is imported by lodash/isEqual?commonjs-external, but could not be resolved – treating it as an external dependency
'lodash/uniqWith' is imported by lodash/uniqWith?commonjs-external, but could not be resolved – treating it as an external dependency
'lodash/pickBy' is imported by lodash/pickBy?commonjs-external, but could not be resolved – treating it as an external dependency
Warnings:

  Message :   Collection list operation response schema should define array property with name: challenge
  Rule    :   ibm-collection-array-property
  Path    :   components.schemas.ChallengePagedResults
  Line    :   9

  Message :   Operation summaries should not have a trailing period
  Rule    :   ibm-summary-sentence-style
  Path    :   paths./challenge.get.summary
  Line    :   40

  Message :   Operation ids must be snake case
  Rule    :   ibm-operationid-casing-convention
  Path    :   paths./challenge.get.operationId
  Line    :   45

  Message :   Response bodies should include an example response
  Rule    :   ibm-success-response-example
  Path    :   paths./challenge.get.responses.200.content.application/json
  Line    :   74

Summary:

  Total number of errors   : 0
  Total number of warnings : 4

  Warnings:
   1  (25%) : Collection list operation response schema should define array property with name
   1  (25%) : Operation summaries should not have a trailing period
   1  (25%) : Operation ids must be snake case
   1  (25%) : Response bodies should include an example response
```

cc @thomasyu888 @brucehoff

### CODE REVIEW

File: libs/synapse/api-description/build/openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  title: Synapse REST API
+  version: 0.1.0
+  license:
+    name: Apache 2.0
+    url: https://github.com/Sage-Bionetworks/sage-monorepo
+  contact:
+    name: Support
+    url: https://github.com/Sage-Bionetworks/sage-monorepo
+  x-logo:
+    url: https://Sage-Bionetworks.github.io/rocc-schemas/logo.png
```
Comment (thomasyu888) [inline libs/synapse/api-description/build/openapi.yaml:29]: Nit:  Should this be different?

File: libs/synapse/api-description/src/components/schemas/ChallengePagedResults.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+title: ChallengePagedResults
+type: object
+properties:
+  totalNumberOfResults:
+    type: integer
+    description: 'The total number of results'
+  # results:
```
Comment (thomasyu888) [inline libs/synapse/api-description/src/components/schemas/ChallengePagedResults.yaml:7]: I think this should exist as outlined here: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/ChallengePagedResults.html

File: libs/synapse/api-description/src/components/schemas/PageMetadata.yaml
Comment (thomasyu888) [inline libs/synapse/api-description/src/components/schemas/PageMetadata.yaml:1]: the Synapse API doesn't have these concepts outlined here.  The paginated results usually follow the `ChallengePagedResults` which is:

```
title: PagedResults
type: object
properties:
  totalNumberOfResults:
    type: integer
    description: "The total number of results"
  results:
    type: array
    description: "The list of results for this page"
    items:
      $ref: '...'
description: "List the Challenges for which the given participant is registered. To be in the returned list the caller must have READ permission on the project associated with the Challenge."

**Review by thomasyu888 — COMMENTED** [review]
This is nice start - see my comments.

I suspect maintaining two sources of truth won't be sustainable.

File: libs/synapse/api-description/build/openapi.yaml (hunk 1 above)
Comment (tschaffter) [reply libs/synapse/api-description/build/openapi.yaml]: I removed info properties that are not strictly required for the sake of keeping the example minimal.

File: libs/synapse/api-description/src/components/schemas/ChallengePagedResults.yaml (hunk 2 above)
Comment (tschaffter) [reply libs/synapse/api-description/src/components/schemas/ChallengePagedResults.yaml]: I restored the property that I had commented out.

File: libs/synapse/api-description/src/components/schemas/PageMetadata.yaml
Comment (tschaffter) [reply libs/synapse/api-description/src/components/schemas/PageMetadata.yaml:1]: I removed the schema `PageMetadata`.

**Pull request merged**